Support mhlo collective ops

okkwon opened the pull request:
Add the support for the following mhlo collective ops:
- mhlo.replica_id
- mhlo.all_gather
- mhlo.all_reduce
- mhlo.reduce_scatter

Since NCCL only supports the splitting and concatenation on dim 0 for all_gather and reduce_scatter, transposes are inserted when the split/concat dimension is not 0.

To make the implementation simple and incremental, several stages are planned as follows:

Stage 1 (The current PR):

It assumes a deterministic order of collective operations with 1:1 mapping from replica_id to rank. This means that there is a single replica group in the mhlo operation and all replicas participate in the collective operation. Since the order is deterministic and all ranks are involved in the communication, we can simply use the default channel for communication. In the MHLO ops, `use_global_device_ids` is set to use the flattened IDs.

Note that the MHLO collective ops have multiple strategies to interpret `replica_groups` attribute, such as `flattened_ids`, `cross_replica`, `cross_partition`, and `cross_replica_and_partition`. (See https://github.com/openxla/stablehlo/blob/main/docs/spec.md#parallel-execution for more details of the strategies.)

Stage 2:

Supports multiple channels. This allows us to have multiple replica groups for collective operations.

Stage 3:

Supports `partition_id` the other strategies: `cross_replica`, `cross_partition`, and `cross_replica_and_partition`.

Stage 4:

Supports `all_to_all` and `collective_permute`. This would need to support the NCCL group markers to support multiple collective ops in parallel since the ops are composite and will need to be lowered into the existing collective ops.

Stage 5:
PJRT integration and model level testing.

okkwon committed:
lower mhlo::ReplicaIdOp to flow ops

Assuming the number of partitions is 1 so that replica_id can be 1:1 mapped to
the rank.

TODO: support partition_id and generalize the mapping of
(replica_id, parition_id) to rank.

okkwon committed:
lower mhlo.all_reduce to flow op

okkwon committed:
lower flow channel.[default|count|rank] to stream ops

okkwon committed:
lower flow.allreduce to stream ops

okkwon committed:
fix out of bound access in checking contiguous dimensions

okkwon committed:
allow -1 for queue_affinity

okkwon committed:
fix the wrong starting value for reduction op enum

okkwon committed:
lower mhlo.all_gather to flow ops

okkwon committed:
lower flow.allgather to stream

okkwon committed:
lower mhlo.reduce_scatter to flow.reduce_scatter

okkwon committed:
lower flow.reduce_scatter to stream ops

okkwon committed:
Add summary and description

benvanik commented on compiler/src/iree/compiler/Dialect/Flow/IR/FlowOps.td:
```diff
@@ -1304,4 +1304,141 @@ def FLOW_DispatchWorkgroupCountFromSetEncodingOp :
+  }];
+}
+
+def FLOW_AllGatherOp : FLOW_Op<"allgather", [
```
lets prefix these ops - dialects are flat namespaces and we want to try to provide structure where we can - here all the collective ops can be prefixed to indicate they are related.

These would become `FLOW_CollectiveAllGatherOp : FLOW_Op<"collective.all_gather">`.
We should also use underscores to separate words. MHLO does this with its ops too (`mhlo.all_gather`) so it'd be good for consistency.

benvanik commented on compiler/src/iree/compiler/Dialect/Flow/IR/FlowOps.td:
```diff
@@ -1304,4 +1304,141 @@ def FLOW_DispatchWorkgroupCountFromSetEncodingOp :
+
+  let arguments = (ins
+    FLOW_CollectiveElementTypeAttr:$element_type,
+    AnyTypeOf<[AnyTensor, AnyMemRef]>:$target,
```
we only accept tensors in flow so you can make this just `FLOW_Tensor`

benvanik commented on compiler/src/iree/compiler/Dialect/Flow/IR/FlowOps.td:
```diff
@@ -1304,4 +1304,141 @@ def FLOW_DispatchWorkgroupCountFromSetEncodingOp :
+  let results = (outs
+    AnyTypeOf<[AnyTensor, AnyMemRef]>:$result
+  );
+  let assemblyFormat = "$element_type `,` operands attr-dict `:` functional-type(operands, results)";
```
For all these please match the existing style of the assemblyFormats in the dialect.

If these are tied ops then you'll need to use the tied op format ala TensorUpdateOp (`custom<ShapedTiedResult>`).

Flow ops must also capture all dimensions. Use the existing ops in the dialect to see how this is done - generally things like `type($update) (`{` $update_dims^ `}`)?` and then capturing dims as args:
```
    FLOW_Tensor:$update,
    FLOW_ShapeDynamicDims:$update_dims,
```

okkwon replied:
Hmm. I am trying to make it work, but it prints out `%2 as tensor<4x2xf32>` instead of `tensor<4x2xf32>`. 

```mlir
%3 = flow.collective.all_gather f32, %2, %1, %channel_default : (tensor<4x2xf32>, tensor<2x2xf32>, !flow.channel) -> %2 as tensor<4x2xf32>
```

This is what I got:
```tablegen
def FLOW_CollectiveAllGatherOp : FLOW_Op<"collective.all_gather", [
  AllTypesMatch<["target", "result"]>,
  DeclareOpInterfaceMethods<Util_TiedOpInterface, [
    "getTiedResult",
    "getTiedResultOperandIndex",
    "getTiedResultOperandIndices",
  ]>,
]> {
  let summary = [{performs all-gather operation}];
  let description = [{It gathers data from all ranks and concatenates them on the 0-th dimension.}];

  let arguments = (ins
    FLOW_CollectiveElementTypeAttr:$element_type,
    FLOW_Tensor:$target,
    FLOW_ShapeDynamicDims:$target_dims,
    FLOW_Tensor:$source,
    FLOW_Channel:$channel,
    OptionalAttr<Util_TiedOpStorageAttr>:$tied_operands
  );
  let results = (outs
    FLOW_Tensor:$result
  );
  let assemblyFormat = [{
    $element_type `,` $target `,` $source `,` $channel `:`
    `(` type($target) `,` type($source) `,` type($channel) `)` `->`
    custom<ShapedTiedResult>(type($result), $target_dims, $tied_operands)
    attr-dict-with-keyword
  }];
  let builders = [
    OpBuilder<(ins
      "CollectiveElementTypeAttr":$element_type,
      "Value":$target,
      "Value":$source,
      "Value":$channel)>,
  ];
}
```

What is missing? I just need to guarantee that the target and the result have the same shape.

okkwon replied:
@benvanik Please take a look at the comment above. Thanks!

okkwon replied:
Okay, it seems like I need to add something like `%target as tensor<2304xf32>` to check the generated output. For input I can use just `tensor<2304xf32>`. The PR got updated.

benvanik commented on compiler/src/iree/compiler/Dialect/Stream/Conversion/FlowToStream/Patterns.cpp:
```diff
@@ -482,6 +482,151 @@ struct ConvertReturnOp : public OpConversionPattern<IREE::Flow::ReturnOp> {
 
+//-------------------------------------------
+// Collective Ops
+//-------------------------------------------
```
these blocks have a standard form that pad out to 80 cols - please match the style already in the files - easiest just to copy and paste these things

```c
// -----------------------------------------------------------------------------
// Collective Ops
// -----------------------------------------------------------------------------
<<WHITESPACE LINE>>
```

benvanik commented on compiler/src/iree/compiler/Dialect/Stream/Conversion/FlowToStream/Patterns.cpp:
```diff
@@ -482,6 +482,151 @@ struct ConvertReturnOp : public OpConversionPattern<IREE::Flow::ReturnOp> {
+    auto shape = op.getSource().getType().cast<ShapedType>();
+    auto collectiveAttr = IREE::Stream::CollectiveAttr::get(
+        op.getContext(), IREE::Stream::CollectiveKind::AllGather,
+        /*CollectiveReductionOp=*/std::nullopt,
```
the syntax for the args here is `/*[arg name]=*/`, and in this case the arg is `reduction` (`mlir::Optional<CollectiveReductionOp> reduction`) so this should be `/*reduction=*/`.

https://google.github.io/styleguide/cppguide.html#Function_Argument_Comments

benvanik commented on compiler/src/iree/compiler/Dialect/Stream/Conversion/FlowToStream/Patterns.cpp:
```diff
@@ -482,6 +482,151 @@ struct ConvertReturnOp : public OpConversionPattern<IREE::Flow::ReturnOp> {
+    auto shape = op.getType().cast<ShapedType>();
+    auto collectiveAttr = IREE::Stream::CollectiveAttr::get(
+        op.getContext(), IREE::Stream::CollectiveKind::AllReduce,
+        static_cast<IREE::Stream::CollectiveReductionOp>(op.getReductionOp()),
```
since you're doing a direct mapping like this we should make sure there's a comment on the enum in FlowBase.td noting that it must exactly match the StreamBase.td equivalent

benvanik commented on compiler/src/iree/compiler/InputConversion/MHLO/MHLOToLinalgOnTensors.cpp:
```diff
@@ -386,8 +389,13 @@ struct ConvertMHLOToLinalgOnTensorsPass
     patterns.insert<GenericTypeConvert>(
         ml_program::GlobalStoreOp::getOperationName(), *typeConverter, context);
-
+    // needed to convert mhlo::ReplicaIDOp
```
style nit: https://google.github.io/styleguide/cppguide.html#Punctuation,_Spelling_and_Grammar

benvanik commented on compiler/src/iree/compiler/InputConversion/MHLO/test/convert_collective_ops.mlir:
```diff
@@ -0,0 +1,167 @@
+  // CHECK-DAG: [[CHANNEL:%.+]] = flow.channel.default : !flow.channel
+  // CHECK-DAG: [[RANK:%.+]] = flow.channel.rank [[CHANNEL]] : index
+  // CHECK-DAG: [[CAST:%.+]] = arith.index_castui [[RANK]] : index to i32
+  // CHECK-DAG: [[TENSOR:%.+]] = tensor.from_elements [[CAST]] : tensor<i32>
```
if real mhlo programs use this we'll want to ensure detensoring is properly handling these (that anything consuming this replica_id gets turned into ops that never end up in tensor types) - fine for now as this is correct but it'll be really inefficient if it ends up sneaking through as a tensor

okkwon replied:
This is a great catch! I wrote a test program, and I found that `linalg-detensorize` didn't optimize the case by default. When I use the pass's sub-option, `aggressive-mode=true`, it detensorizes correctly.

Here is the input and the detensorized output:
```mlir
func.func @replica_id_dynamic_slice(%in : tensor<1024xf32>) -> tensor<512xf32> {
  %id = mhlo.replica_id : tensor<ui32>
  %c2 = mhlo.constant dense<2> : tensor<ui32>
  %rem = mhlo.remainder %id, %c2 : tensor<ui32>
  %c512 = mhlo.constant dense<512> : tensor<ui32>
  %mul = mhlo.multiply %rem, %c512 : tensor<ui32>
  %slice = "mhlo.dynamic_slice"(%in, %mul) {slice_sizes = dense<512> : tensor<1xi64>} : (tensor<1024xf32>, tensor<ui32>) -> tensor<512xf32> 
  return %slice : tensor<512xf32>
}

// -----// IR Dump After FusionOfTensorOps (iree-flow-fusion-of-tensor-ops) //----- //
func.func @replica_id_dynamic_slice(%arg0: !hal.buffer_view) -> !hal.buffer_view attributes {iree.abi.stub} {
  %c2_i32 = arith.constant 2 : i32
  %c512_i32 = arith.constant 512 : i32
  %c0 = arith.constant 0 : index
  %c512 = arith.constant 512 : index
  %0 = hal.tensor.import %arg0 : !hal.buffer_view -> tensor<1024xf32>
  %channel_default = flow.channel.default : !flow.channel
  %channel_rank = flow.channel.rank %channel_default : index
  %1 = arith.index_castui %channel_rank : index to i32
  %from_elements = tensor.from_elements %1 : tensor<i32>
  %2 = tensor.empty() : tensor<i32>
  %3 = linalg.generic {indexing_maps = [affine_map<() -> ()>, affine_map<() -> ()>], iterator_types = []} ins(%from_elements : tensor<i32>) outs(%2 : tensor<i32>) {
  ^bb0(%in: i32, %out: i32):
    %8 = arith.remui %in, %c2_i32 : i32
    %9 = arith.muli %8, %c512_i32 : i32
    linalg.yield %9 : i32
  } -> tensor<i32>
  %extracted = tensor.extract %3[] : tensor<i32>
  %4 = arith.index_castui %extracted : i32 to index
  %5 = arith.maxsi %4, %c0 : index
  %6 = arith.minsi %5, %c512 : index
  %extracted_slice = tensor.extract_slice %0[%6] [512] [1] : tensor<1024xf32> to tensor<512xf32>
  %7 = hal.tensor.export %extracted_slice : tensor<512xf32> -> !hal.buffer_view
  return %7 : !hal.buffer_view
}

// -----// IR Dump After LinalgDetensorize (linalg-detensorize) //----- //
func.func @replica_id_dynamic_slice(%arg0: !hal.buffer_view) -> !hal.buffer_view attributes {iree.abi.stub} {
  %c2_i32 = arith.constant 2 : i32
  %c512_i32 = arith.constant 512 : i32
  %c0 = arith.constant 0 : index
  %c512 = arith.constant 512 : index
  %0 = hal.tensor.import %arg0 : !hal.buffer_view -> tensor<1024xf32>
  %channel_default = flow.channel.default : !flow.channel
  %channel_rank = flow.channel.rank %channel_default : index
  %1 = arith.index_castui %channel_rank : index to i32
  %2 = arith.remui %1, %c2_i32 : i32
  %3 = arith.muli %2, %c512_i32 : i32
  %4 = arith.index_castui %3 : i32 to index
  %5 = arith.maxsi %4, %c0 : index
  %6 = arith.minsi %5, %c512 : index
  %extracted_slice = tensor.extract_slice %0[%6] [512] [1] : tensor<1024xf32> to tensor<512xf32>
  %7 = hal.tensor.export %extracted_slice : tensor<512xf32> -> !hal.buffer_view
  return %7 : !hal.buffer_view
}
```

okkwon replied:
@MaheshRavishankar do you happen to know why the default detensorizer fails? Also, how can I enable the sub-option in the command line? Thanks!

benvanik replied:
awesome, thanks for trying this! would indeed be really good to get it turned on (has been needed for a very long time!), or at least ensure you're doing your testing with it on until it can be on by default so you don't end up measuring the slow stuff :)

benvanik commented on runtime/src/iree/hal/command_buffer.h:
```diff
@@ -297,7 +297,7 @@ typedef uint8_t iree_hal_collective_kind_t;
 // Specifies the reduction operator of a collective reduction operation.
 enum iree_hal_collective_reduction_e {
   // Specifies that the reduction operation computes a sum (addition).
-  IREE_HAL_COLLECTIVE_REDUCTION_SUM = 0,
```
good catch - make an `IREE_HAL_COLLECTIVE_REDUCTION_NONE = 0` to match the HALBase.td enum instead of starting at 1

benvanik commented on runtime/src/iree/hal/drivers/cuda/cuda_device.c:
```diff
@@ -301,7 +301,8 @@ static iree_status_t iree_hal_cuda_device_create_channel(
   // compiler so that it can emit the right rank math.
   int requested_count = iree_math_count_ones_u64(queue_affinity);
-  if (requested_count != 1) {
+  // FIXME: queue affinity is not set yet correctly, so we have all bits set.
```
this isn't going to work when we have multiple queues in CUDA - if you need this to unblock yourself then please file an issue and tag it here with `// TODO(#issue): properly assign affinity in the compiler.`

benvanik commented on compiler/src/iree/compiler/Dialect/Flow/IR/FlowOps.cpp:
```diff
@@ -1545,6 +1545,78 @@ void populateTensorSliceOpWithDispatchTensorOpFoldingPatterns(
+//===----------------------------------------------------------------------===//
+// flow.channel.default
+//===----------------------------------------------------------------------===//
+void ChannelDefaultOp::getAsmResultNames(
```
vertical whitespace here and elsewhere to match the existing style in these files

benvanik commented on compiler/src/iree/compiler/InputConversion/MHLO/ConvertCollectiveOps.cpp:
```diff
@@ -0,0 +1,411 @@
+}  // namespace
+
+/// Converts mhlo.replica_id to flow.channel.default + flow.channel.rank.
+/// TODO: this assumes that there is no partition so that there is a 1:1 mapping
```
use either issues or area owner for todos, like `// TODO(okkwon):` so that people in the future (or yourself) can know who holds context on the work to be done.

https://google.github.io/styleguide/cppguide.html#TODO_Comments

benvanik commented on compiler/src/iree/compiler/InputConversion/MHLO/ConvertCollectiveOps.cpp:
```diff
@@ -0,0 +1,411 @@
+    SmallVector<int64_t> gatherResultShape(resultType.getShape());
+    auto elemType = resultType.getElementType();
+
+    // Get the element type
```
https://google.github.io/styleguide/cppguide.html#Punctuation,_Spelling_and_Grammar

benvanik commented on compiler/src/iree/compiler/InputConversion/MHLO/ConvertCollectiveOps.cpp:
```diff
@@ -0,0 +1,411 @@
+      permutationAttr = rewriter.getI64VectorAttr(permutation);
+      std::swap(gatherInputShape[0], gatherInputShape[allGatherDim]);
+      std::swap(gatherResultShape[0], gatherResultShape[allGatherDim]);
+      // transpose the input
```
https://google.github.io/styleguide/cppguide.html#Punctuation,_Spelling_and_Grammar (etc etc, please correct all)

benvanik commented on compiler/src/iree/compiler/InputConversion/MHLO/ConvertCollectiveOps.cpp:
```diff
@@ -0,0 +1,411 @@
+
+    if (op1.getNumResults() != 1 ||
+        !op1.hasTrait<::mlir::OpTrait::Elementwise>())
+      return rewriter.notifyMatchFailure(op, "must have elementwise trait");
```
any multi-line ifs should be wrapped with braces (here and above/below) - llvm is pretty lax on this but in the IREE codebase we follow the google guideline there: https://google.github.io/styleguide/cppguide.html#Conditionals

the easiest rule to remember is to always add braces - it's always safer and unless hyper optimizing to save one vertical whitespace line is not any less readable. part of this is because avoiding the braces makes refactoring and automatic reformatting much more complex: a line that may have just fit because the error message string had one word instead of another may suddenly not with a global find/replace, and then all places either end up inconsistent or need to be fixed. or if someone comes along and decides they need to do one more thing instead of a diff of 1 line they now need to change 3 to add the braces. It's the same reason having trailing commas on initializer lists is useful.

benvanik commented on compiler/src/iree/compiler/InputConversion/MHLO/ConvertCollectiveOps.cpp:
```diff
@@ -0,0 +1,411 @@
+
+    auto resultType = op.getResult().getType().cast<RankedTensorType>();
+    SmallVector<int64_t> gatherResultShape(resultType.getShape());
+    auto elemType = resultType.getElementType();
```
there's a lot of duplication in these patterns - given that there will be more we should factor this out - especially the things like channel selection as we'll want that logic to be shared across all of the ops. generally you can do this by having a static helper function above that takes the `Operation *` and an `OpBuilder &` (the rewriter), does whatever logic you need (here just creating the default channel), and returns it. Same for the things like type checks and extractions, creating the result output, etc.

okkwon replied:
I added a helper function to get the collective element type attribute to clean up a little more. Regarding the channel creation, the clean up needs to be done later when we get each channel from the replica groups since this would form a loop in the code. Thanks!

okkwon committed:
prefix flow collective ops with collective

Also allreduce and allgather are renamed to all_reduce and all_gather,
respectively.

okkwon committed:
use FLOW_Tensor

okkwon committed:
style fixes and more clear comments

okkwon commented on compiler/src/iree/compiler/Dialect/Flow/Transforms/CollapseReductionDims.cpp:
```diff
@@ -31,7 +31,8 @@ static bool hasContiguousDims(AffineMap map, ArrayRef<unsigned> dims) {
     }
     // Check that the following dimensions are match the order of `dims`
     for (unsigned j = 1, numDims = dims.size(); j < numDims; j++) {
-      if (map.getDimPosition(i + j) != dims[j]) {
```
Filed https://github.com/iree-org/iree/issues/12211

okkwon committed:
use a helper function to get the collective element type attr

okkwon committed:
match the tied op convention used in Flow
